src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -1363,26 +1333,33 @@ def forward(
         else:
             raise ValueError("You have to specify either decoder_input_ids or decoder_inputs_embeds")
 
-        # past_key_values_length
-        past_key_values_length = past_key_values[0][0].shape[2] if past_key_values is not None else 0
-
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input_ids)
 
-        if self._use_flash_attention_2:
-            # 2d mask is passed through the layers
-            attention_mask = attention_mask if (attention_mask is not None and 0 in attention_mask) else None
-        elif self._use_sdpa and head_mask is None and not output_attentions:
-            # output_attentions=True & head_mask can not be supported when using SDPA.
-            attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
-                attention_mask, input_shape, inputs_embeds, past_key_values_length
-            )
-        else:
-            # 4d mask is passed through the layers
-            attention_mask = _prepare_4d_causal_attention_mask(
-                attention_mask, input_shape, inputs_embeds, past_key_values_length
+        return_legacy_cache = False
+        return_self_attention_cache = False
+        if use_cache or past_key_values is not None:
+            if isinstance(past_key_values, Cache):
+                return_self_attention_cache = True
+                past_key_values = EncoderDecoderCache(past_key_values, DynamicCache())
+            elif not isinstance(past_key_values, EncoderDecoderCache):
+                return_legacy_cache = True
+                past_key_values = EncoderDecoderCache.from_legacy_cache(past_key_values)
```

Comment:
Ohh nice trick. Only downside for me is that this is not super explicit. 
The idea is that past_key_values here is always the decoder's PKV. NIce, but should be done in the generate! 

Because EncoderDecoderCache should be a Cache no? 

That is the trickiest indeed !

Reply:
Regarding converting a self-attention cache to an encoder-decoder cache, this is a bit tricky to do in `.generate`.

Consider the case when we use a decoder-only Whisper model as the assistant to an encoder-decoder model. The assistant model has `config.is_encoder_decoder=False`, since it's decoder-only, which mean all the checks for encoder-decoder fail.

This didn't matter with the old `tuple` cache design, since we constructed the past key-values in the `.forward` method, so could create a tuple of self-attention and cross-attention cache.

However, now that we create the cache in `.generate`, we need to instantiate an **encoder-decoder cache** for a **decoder-only** model. Since we don't hit any of the triggers for instantiating an encoder-cache in `.generate`, the only options would be to add a specific exception for Whisper:

```python
if self.config.is_encoder_decoder or isinstance(self, WhisperForCausalLM):
```

This is quite a big edge case, for which I don't think it's worth convoluting `.generate` with model-specific logic, and instead handle it as we did before: directly in the `.forward`.

Otherwise, all we're doing is guarding against the case where a user passes a **tuple** of `past_key_values`, in which case we convert it to the correct class `EncoderDecoderCache` in the `.forward`.

This is one-to-one the same as LLaMA, where if a user passes a tuple of `past_key_values`, we convert it to `DynamicCache` in the `.forward`:
https://github.com/huggingface/transformers/blob/3f93fd06949f9eae58e50fd0c9b8e60be82643bc/src/transformers/models/llama/modeling_llama.py#L934-L936